Handle announcement only streams.

roberthoenig opened the pull request:
Fixes #2711.

roberthoenig commented:
Whoops, forgot to post the demo gifs.

### Normal user
![bug7](https://user-images.githubusercontent.com/7950151/42047078-9b20ef72-7aff-11e8-84f1-8b790aa23235.gif)

### Admin user
![bug8](https://user-images.githubusercontent.com/7950151/42047085-9f1e6604-7aff-11e8-9c52-c68f77bff5a3.gif)

borisyankov commented on src/message/AnnouncementOnly.js:
```diff
@@ -0,0 +1,37 @@
+  }
+}
+
+export default connect()(AnnouncementOnly);
```
You don't need `connect`.

borisyankov commented on src/streams/streamSelectors.js:
```diff
@@ -0,0 +1,15 @@
+import { getStreams } from '../directSelectors';
+import { isStreamOrTopicNarrow } from '../utils/narrow';
+
+export const getIsActiveStreamAnnouncementOnly = (narrow: Narrow) =>
```
Feel free to create a new commit for that in this PR.

roberthoenig commented:
Updated:
* Rebased
* Fixed the FloxFixMe
* Removed the unnecessary connect()
* moved getIsActiveStreamAnnouncementOnly to subscriptionSelectors and avoided creating a separate streamSelectors. I won't do the refactor of extracting the selectors in subscriptionSelectors that are actually stream selectors here, but can do that in another PR.

roberthoenig commented:
Rebased.

roberthoenig commented:
Updated:

* added the i18n string
* extracted the `is_announcement_only` flow type into a new commit
* aligned the import statements in ComposeBox.*.js
* Swapped `styles: Extract disabledCompose* styles.` and `Handle narrow to announcement only stream.`

gnprice commented:
Merged -- thanks @roberthoenig !

The revisions look great; pulling the corrected type into its own commit was a particularly good idea.

A couple of small issues which I fixed before merging:
* That first commit correcting the type actually caused type errors on its own -- it needed to go together with the updates in `nullObjects.js`. (I used the same technique Tim demonstrated last week to pull those changes forward into the first commit.)
* The last commit message still explained why we showed this message in preference to the "not subscribed" message when both applied; it hadn't been updated for the revised UI choice.
